From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Sat, 16 Sep 2023 21:34:57 +0400
Subject: [PATCH] Do-not-send-chunk-if-it-was-already-sent


diff --git a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
index 239475187fdf0d05823b31e2068651559d3497e5..d4ded2324d4c1c2ec3a105fa24facef1d5d57cdc 100644
--- a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
@@ -421,6 +421,10 @@ public class RegionizedPlayerChunkLoader {
         }
 
         private void sendChunk(final int chunkX, final int chunkZ) {
+            if (this.sentChunks.contains(CoordinateUtils.getChunkKey(chunkX, chunkZ))){
+                return;
+            }
+
             if (this.sentChunks.add(CoordinateUtils.getChunkKey(chunkX, chunkZ))) {
                 PlayerChunkSender.sendChunk(this.player.connection, this.world, this.world.getChunkIfLoaded(chunkX, chunkZ));
                 return;
