From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Thu, 4 Jan 2024 14:49:50 +0400
Subject: [PATCH] Pufferfish-Optimize-entity-coordinate-key


diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index 08d282400644be5949f28eb879b333cc221da2cc..bada1ff427be6d5892a3eb9b3af0a3d9e79cd2ee 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -215,7 +215,7 @@ public final class MCUtil {
     }
 
     public static long getCoordinateKey(final Entity entity) {
-        return ((long)(MCUtil.fastFloor(entity.getZ()) >> 4) << 32) | ((MCUtil.fastFloor(entity.getX()) >> 4) & 0xFFFFFFFFL);
+        return ((long)(entity.blockPosition().getZ() >> 4) << 32) | ((entity.blockPosition().getX() >> 4) & 0xFFFFFFFFL);
     }
 
     public static long getCoordinateKey(final ChunkPos pair) {
