From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AltronMaxX <max06112004@gmail.com>
Date: Tue, 6 Aug 2024 16:19:01 +0400
Subject: [PATCH] Add-chance-config-for-piglin-spawn-in-nether-portal


diff --git a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
index b52f7bcc17fba6a3a4376a4bfea17430be9c4a8e..553c02a53b9aed6c851305f1e885934677fa5daa 100644
--- a/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
+++ b/src/main/java/net/edenor/foldenor/config/FoldenorConfig.java
@@ -45,6 +45,8 @@ public class FoldenorConfig {
 
     public static boolean useAlternateKeepAlive = false;
 
+    public static int piglinSpawnChancePersentInPortal = 100;
+
     public static void init(File configFile) {
         init(configFile, true);
     }
@@ -114,6 +116,8 @@ public class FoldenorConfig {
                 "the check in a way that still appears vanilla. This should",
                 "be left enabled on most servers, but is provided as a",
                 "configuration option if the vanilla deviation is undesirable.");
+        piglinSpawnChancePersentInPortal = getInt("optimizations.piglin-spawn-chance-persent-in-portal", 100,
+            "Reduces piglin spawn in portal, by reducing change to spawn");
     }
 
     private static void readDynamicActivationOfBrains() throws IOException {
diff --git a/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java b/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
index eba62f1336697157da94c7dcde389cc2e929bb0b..918b50e1f17fa2b49e5e7ef48182538ea12aae02 100644
--- a/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -4,6 +4,8 @@ import com.mojang.logging.LogUtils;
 import com.mojang.serialization.MapCodec;
 import java.util.Optional;
 import javax.annotation.Nullable;
+
+import net.edenor.foldenor.config.FoldenorConfig;
 import net.minecraft.BlockUtil;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -77,7 +79,9 @@ public class NetherPortalBlock extends Block implements Portal {
 
     @Override
     protected void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
-        if (world.spigotConfig.enableZombiePigmenPortalSpawns && world.dimensionType().natural() && world.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && random.nextInt(2000) < world.getDifficulty().getId()) { // Spigot
+        if (world.spigotConfig.enableZombiePigmenPortalSpawns && world.dimensionType().natural()
+            && world.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && random.nextInt(2000) < world.getDifficulty().getId()
+            && random.nextInt(100) <= FoldenorConfig.piglinSpawnChancePersentInPortal) { // Spigot // Foldenor
             while (world.getBlockState(pos).is((Block) this)) {
                 pos = pos.below();
             }
